### PR TITLE
core: Add OFI_GETINFO_HIDDEN flag for fi_getinfo()

### PR DESCRIPTION
**Pull request by rwespetal**
Store the providers that are excluded on a separate hidden list that can
be queried via fi_getinfo() with the OFI_GETINFO_HIDDEN flag. Providers
can use this flag to find another core provider that is excluded.

Signed-off-by: Robert Wespetal <wesper@amazon.com>

### CODE REVIEW

**Comment by rwespetal** [comment]
aws:bot:retest

File: src/fabric.c [hunk 1]
```diff
@@ -254,11 +255,13 @@ static void ofi_filter_info(struct fi_info **info)
...
 }
 
-static struct ofi_prov *ofi_getprov(const char *prov_name, size_t len)
+static struct ofi_prov *ofi_getprov(const char *prov_name, size_t len,
+                                    bool hidden)
 {
 	struct ofi_prov *prov;
 
-	for (prov = prov_head; prov; prov = prov->next) {
+	for (prov = hidden ? prov_head_hidden : prov_head;
```
Comment (rajachan) [inline src/fabric.c]: I assumed you were going to have a separate list of providers that qualified to be hidden when excluded, but I like this even better.

**Comment by shefty** [comment]
The flag is good.  I was envisioning that when the hidden flag was set it would find both hidden and visible providers.  This way, the underlying provider wouldn't need to know if the core provider was hidden or not.

It may be simpler in that case to extend struct ofi_prov with a hidden attribute and maintain a single provider list.  When looping over fi_getinfo(), just skip any hidden provider unless the hidden flag was also specified.

**Comment by rwespetal** [comment]
Sure, that makes sense. Let me update the patch.

**Comment by rwespetal** [comment]
I updated the patch, let me know if you need additional changes.

**Pull request merged**